File: extensions/soomla-cocos2dx-core/development/submodules/soomla-android-core/src/com/soomla/SoomlaUtils.java
/*
 * Copyright (C) 2012-2014 Soomla Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.soomla;

import android.provider.Settings;
import android.text.TextUtils;
import android.util.Log;

import com.soomla.data.KeyValueStorage;

import java.util.Random;

/**
 * This class provides Log functions that output debug, warning, or error messages.
 */
public class SoomlaUtils {

    /**
     * Creates Log Debug message according to given tag and message.
     *
     * @param tag the name of the class whose instance called this function
     * @param message debug message to output to log
     */
    public static void LogDebug(String tag, String message) {
        if (SoomlaConfig.logDebug) {
            Log.d(tag, message);
        }
    }

    /**
     * Creates Log Warning message according to given tag and message.
     *
     * @param tag the name of the class whose instance called this function
     * @param message warning message to output to log
     */
    public static void LogWarning(String tag, String message) {
        Log.w(tag, message);
    }

    /**
     * Creates Log Error message according to given tag and message.
     *
     * @param tag the name of the class whose instance called this function
     * @param message error message to output to log
     */
    public static void LogError(String tag, String message) {
        Log.e(tag, message);
    }

    /**
     * Retrieves Android device Id.
     *
     * @return androidId which is the id of the device being used
     */
    public static String deviceId() {
        String androidId = Settings.Secure.getString(SoomlaApp.getAppContext().getContentResolver(),
                Settings.Secure.ANDROID_ID);
        if (androidId == null) {
            // This is a fallback in case the device id cannot be retrieved on the device
            // (happened on some devices !)
            SoomlaUtils.LogError("SOOMLA ObscuredSharedPreferences",
                    "Couldn't fetch ANDROID_ID. Using generated SOOMLA id.");
            androidId = generateSoomlaId();
        }

        return androidId;
    }

    /**
     * Retrieves the class name (to be used for serialization/deserialization
     * for the passed instance
     * @param target The instance to get the class name for
     * @return the target's class name
     */
    public static String getClassName(Object target) {
        return target.getClass().getSimpleName();
    }

	private static String generateSoomlaId() {
		String generatedId = KeyValueStorage.getValue(DB_KEY_SOOMLA_GENERATED_ID);
		if (TextUtils.isEmpty(generatedId)) {
			generatedId = "SOOMLA_ID_a" + String.format("%05d", new Random().nextInt(100000)) + String.format("%05d", new Random().nextInt(100000));
			KeyValueStorage.setValue(DB_KEY_SOOMLA_GENERATED_ID, generatedId);
		}
		return generatedId;
	}


    /** Private Members **/

    private static String TAG = "SOOMLA SoomlaUtils"; //used for Log messages

	private static final String DB_KEY_SOOMLA_GENERATED_ID = SoomlaConfig.DB_KEY_PREFIX + "generatedid";
}

File: extensions/soomla-cocos2dx-core/development/submodules/soomla-ios-core/SoomlaiOSCore/SoomlaUtils.m
/*
 Copyright (C) 2012-2014 Soomla Inc.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

#import "SoomlaUtils.h"
#import "SoomlaConfig.h"
#import <UIKit/UIKit.h>

@implementation SoomlaUtils

static NSString* TAG = @"SOOMLA SoomlaUtils";

static NSString *const SOOMLA_DEVICE_KEY = @"soomlaDeviceId";
static NSString *const SOOMLA_GENERATED_KEY = @"soomlaGeneratedId";

+ (void)LogDebug:(NSString*)tag withMessage:(NSString*)msg {
    if (DEBUG_LOG) {
        NSLog(@"[Debug] %@: %@", tag, msg);
    }
}

+ (void)LogError:(NSString*)tag withMessage:(NSString*)msg {
    NSLog(@"[*** ERROR ***] %@: %@", tag, msg);
}

+ (NSString*)deviceIdPreferVendor {
    NSString *soomlaDeviceId = [[NSUserDefaults standardUserDefaults] stringForKey:SOOMLA_DEVICE_KEY];
    if (!soomlaDeviceId) {
        soomlaDeviceId = [SoomlaUtils vendorId];
        [[NSUserDefaults standardUserDefaults] setObject:soomlaDeviceId forKey:SOOMLA_DEVICE_KEY];
    }
    return soomlaDeviceId;
}

/* We check for UDID_SOOMLA to support devices with older versions of ios-store */
+ (NSString*)deviceId {
    NSString* udid = [[NSUserDefaults standardUserDefaults] stringForKey:@"UDID_SOOMLA"];
    if (!udid || [udid length] == 0) {
	return [self deviceIdPreferVendor];
    }
    return udid;
}

+ (NSString*)vendorId {
    NSString *vendorId;
    if ([[UIDevice currentDevice] respondsToSelector:@selector(identifierForVendor)]) {
        vendorId = [[[UIDevice currentDevice] identifierForVendor] UUIDString];
    } else {
        vendorId = [SoomlaUtils generateSoomlaId];
    }
    return vendorId;
}

+ (NSString*)generateSoomlaId {
    NSString *soomlaGeneratedId = [[NSUserDefaults standardUserDefaults] stringForKey:SOOMLA_GENERATED_KEY];
    if (!soomlaGeneratedId) {
        soomlaGeneratedId = [NSString stringWithFormat:@"SOOMLA_ID_i%05d%05d", arc4random_uniform(100000), arc4random_uniform(100000)];
        [[NSUserDefaults standardUserDefaults] setObject:soomlaGeneratedId forKey:SOOMLA_GENERATED_KEY];
    }
    return soomlaGeneratedId;
}


+ (NSString*)keyFromSecret:(NSString*)secret {
    return [secret stringByAppendingString:[SoomlaUtils deviceId]];
}

+ (NSMutableDictionary*)jsonStringToDict:(NSString*)str {
    NSError* error = NULL;
    NSMutableDictionary *dict =
    [NSJSONSerialization JSONObjectWithData: [str dataUsingEncoding:NSUTF8StringEncoding]
                                    options: NSJSONReadingMutableContainers
                                      error: &error];
    if (error) {
        LogError(TAG, ([NSString stringWithFormat:@"There was a problem parsing the given JSON string: %@ error: %@", str, [error localizedDescription]]));

        return NULL;
    }

    return dict;
}

+ (NSMutableArray*)jsonStringToArray:(NSString*)str {
    NSError* error = NULL;
    NSMutableArray *arr =
    [NSJSONSerialization JSONObjectWithData: [str dataUsingEncoding:NSUTF8StringEncoding]
                                    options: NSJSONReadingMutableContainers
                                      error: &error];
    if (error) {
        LogError(TAG, ([NSString stringWithFormat:@"There was a problem parsing the given JSON string: %@ error: %@", str, [error localizedDescription]]));

        return NULL;
    }

    return arr;
}

+ (NSString*)dictToJsonString:(NSDictionary*)dict {
    NSError *error;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:dict
                                                       options:NSJSONWritingPrettyPrinted
                                                         error:&error];

    if (! jsonData) {
        LogError(TAG, ([NSString stringWithFormat:@"There was a problem parsing the given NSDictionary. error: %@", [error localizedDescription]]));

        return NULL;
    }

    return [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
}

+ (NSString*)arrayToJsonString:(NSArray*)arr {
    NSError *error;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:arr
                                                       options:NSJSONWritingPrettyPrinted
                                                         error:&error];

    if (! jsonData) {
        LogError(TAG, ([NSString stringWithFormat:@"There was a problem parsing the given NSArray. error: %@", [error localizedDescription]]));

        return NULL;
    }

    return [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
}

+ (NSString *) applicationDirectory
{
    static NSString* appDir = nil;

    if (appDir == nil) {
        NSArray *paths = NSSearchPathForDirectoriesInDomains(NSApplicationSupportDirectory, NSUserDomainMask, YES);
        if ([paths count] == 0)
        {
            // *** creation and return of error object omitted for space
            return nil;
        }

        NSString *basePath = paths[0];
        NSError *error;

        NSFileManager *fManager = [NSFileManager defaultManager];
        if (![fManager fileExistsAtPath:basePath]) {
            if (![fManager createDirectoryAtPath:basePath
                     withIntermediateDirectories:YES
                                      attributes:nil
                                           error:&error])
            {
                LogError(TAG, ([NSString stringWithFormat:@"Create directory error: %@", error]));
                return nil;
            }
        }
        appDir = [basePath copy];
    }
    return appDir;
}

+ (BOOL)addSkipBackupAttributeToItemAtURL:(NSURL *)URL
{
    assert([[NSFileManager defaultManager] fileExistsAtPath: [URL path]]);

    NSError *error = nil;
    BOOL success = [URL setResourceValue: [NSNumber numberWithBool: YES]
                                  forKey: NSURLIsExcludedFromBackupKey error: &error];
    if(!success){
        NSLog(@"Error excluding %@ from backup %@", [URL lastPathComponent], error);
    }
    return success;
}

+ (void)setLoggingEnabled:(BOOL)logEnabled {
    DEBUG_LOG = logEnabled;
}

+ (NSString *)getClassName:(id)target {
    return NSStringFromClass([target class]);
}

+ (BOOL)isEmpty:(NSString *)target {
    return (!target || target.length == 0);
}

@end
